lib/charms/observability_libs/v1/cert_handler.py

```diff
@@ -355,6 +360,27 @@ def __init__(
             self._on_upgrade_charm,
         )
 
+        if refresh_events:
+            for ev in refresh_events:
+                self.framework.observe(ev, self._on_refresh_event)
+
+    def _on_refresh_event(self, _):
+        """Replace the latest current CSR with a new one if there are any CSR changes.
+
+        The following CSR changes will trigger a certificate renewal: DNS SANs, IP SANs, subject, and private key changes.
```

Comment:
ok, I think I see now, thanks for the clarification! do you think this behaviour would ever be good to have outside of the refresh event flow?
I mean, will the charm ever want to do something like: `self.cert_handler.refresh_csr()`?

Reply:
if yes, we can pull this logic out to a separate (public?) method...

Reply:
> ok, I think I see now, thanks for the clarification! do you think this behaviour would ever be good to have outside of the refresh event flow? I mean, will the charm ever want to do something like: `self.cert_handler.refresh_csr()`?

Its possible, but do we need it? When certs approach expiry, for example, CertHandler would renew certs on its own without the charm's intervention. So, I think CertHandler is there to handle these sorts of interactions leaving out the need for a charm to manually invoke sth.

Reply:
FYI
Comparing the entire CSR object will not work because the object would contain a UID that will always be different (Thanks to UTs for discovering this)
I refactored things to detect if renewal is needed based on SANs only while keeping the current SANs in a `StoredState`.